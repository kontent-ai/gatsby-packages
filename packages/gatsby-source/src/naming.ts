import { PluginNamingConfiguration } from './types';
import { ItemsSchemaTemplate, TaxonomiesSchemaTemplate, TypesSchemaTemplate } from "./schemaTemplates";

const CONNECTOR = '_';
const SYSTEM_IDENTIFIER = 'system';
const ELEMENT_IDENTIFIER = 'element';
const VALUE_IDENTIFIER = 'value';
const TAXONOMY_TERM_IDENTIFIER = 'term';
const ELEMENT_OPTION_IDENTIFIER = 'option';
const MULTI_ELEMENT_IDENTIFIER = `${ELEMENT_IDENTIFIER}s`;
const LANGUAGE_LINK_EXTENSION_IDENTIFIER = 'language_link';
const ITEM_IDENTIFIER = 'item';
const TAXONOMY_IDENTIFIER = 'taxonomy';
const TYPE_IDENTIFIER = 'type';
const CACHE_IDENTIFIER = 'cache';

const RICH_TEXT_ELEMENT_TYPE_NAME = 'rich_text';

const PREFERRED_LANGUAGE_IDENTIFIER = 'preferred_language';

const defaultPluginNamingConfiguration: PluginNamingConfiguration = {
  prefix: `kontent`,
};

/**
 * Retrieve ID string for Gatsby's CreateNodeId method for specified item language variant.
 * @param id  Item `system.id` because codename is changeable.
 * @param preferredLanguage Preferred language of the language variant.
 * @param config Optional parameter with extra configuration.
 */
const getKontentItemNodeStringForId = (
  id: string,
  preferredLanguage: string,
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}${CONNECTOR}${preferredLanguage}${CONNECTOR}${id}`;

const getKontentItemNodeTypeName = (
  type: string,
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}${CONNECTOR}${type}`;

const getKontentItemInterfaceName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string => `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}`;

const getKontentItemSystemElementTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}${CONNECTOR}${SYSTEM_IDENTIFIER}`;

const getKontentItemElementTypeNameByType = (
  type: string,
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}${CONNECTOR}${type}${CONNECTOR}${ELEMENT_IDENTIFIER}${CONNECTOR}${VALUE_IDENTIFIER}`;

const getKontentItemElementValueTypeNameByType = (
  type: string,
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}${CONNECTOR}${type}${CONNECTOR}${ELEMENT_IDENTIFIER}`;

const getKontentItemElementsSchemaTypeName = (
  type: string,
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}${CONNECTOR}${type}${CONNECTOR}${MULTI_ELEMENT_IDENTIFIER}`;

const getKontentItemLanguageLinkExtensionName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}${CONNECTOR}${LANGUAGE_LINK_EXTENSION_IDENTIFIER}`;

const getKontentTaxonomyNodeStringForCodeName = (
  codename: string,
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${TAXONOMY_IDENTIFIER}${CONNECTOR}${codename}`;

const getKontentTaxonomyTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string => `${config.prefix}${CONNECTOR}${TAXONOMY_IDENTIFIER}`;

const getKontentTaxonomySystemElementTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${TAXONOMY_IDENTIFIER}${CONNECTOR}${SYSTEM_IDENTIFIER}`;

const getKontentTaxonomyTermTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${TAXONOMY_IDENTIFIER}${CONNECTOR}${TAXONOMY_TERM_IDENTIFIER}`;

const getKontentTypeNodeStringForCodeName = (
  codename: string,
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${TYPE_IDENTIFIER}${CONNECTOR}${codename}`;

const getKontentTypeTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string => `${config.prefix}${CONNECTOR}${TYPE_IDENTIFIER}`;

const getKontentTypeSystemElementTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${TYPE_IDENTIFIER}${CONNECTOR}${SYSTEM_IDENTIFIER}`;

const getKontentTypeElementTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${TYPE_IDENTIFIER}${CONNECTOR}${ELEMENT_IDENTIFIER}`;

const getKontentTypeElementOptionTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${getKontentTypeElementTypeName(
    config,
  )}${CONNECTOR}${ELEMENT_OPTION_IDENTIFIER}`;

const getKontentTypesCacheKey = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${TYPE_IDENTIFIER}${CONNECTOR}${CACHE_IDENTIFIER}`;

const getKontentItemsSchemaNamingConfiguration = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string => {
  return (
    ItemsSchemaTemplate
      .replace(
        /__KONTENT_ITEM_INTERFACE__/g,
        getKontentItemInterfaceName(config),
      )
      .replace(
        /__KONTENT_ITEM_SYSTEM_TYPE__/g,
        getKontentItemSystemElementTypeName(config),
      )
      // elements
      .replace(
        /__KONTENT_ITEM_TEXT_ELEMENT__/g,
        getKontentItemElementTypeNameByType('text', config),
      )
      .replace(
        /__KONTENT_ITEM_RICH_TEXT_ELEMENT__/g,
        getKontentItemElementTypeNameByType(RICH_TEXT_ELEMENT_TYPE_NAME, config),
      )
      .replace(
        /__KONTENT_ITEM_NUMBER_ELEMENT__/g,
        getKontentItemElementTypeNameByType('number', config),
      )
      .replace(
        /__KONTENT_ITEM_MULTIPLE_CHOICE_ELEMENT__/g,
        getKontentItemElementTypeNameByType('multiple_choice', config),
      )
      .replace(
        /__KONTENT_ITEM_DATE_TIME_ELEMENT__/g,
        getKontentItemElementTypeNameByType('date_time', config),
      )
      .replace(
        /__KONTENT_ITEM_ASSET_ELEMENT__/g,
        getKontentItemElementTypeNameByType('asset', config),
      )
      .replace(
        /__KONTENT_ITEM_MODULAR_CONTENT_ELEMENT__/g,
        getKontentItemElementTypeNameByType('modular_content', config),
      )
      .replace(
        /__KONTENT_ITEM_CUSTOM_ELEMENT_ELEMENT__/g,
        getKontentItemElementTypeNameByType('custom', config),
      )
      .replace(
        /__KONTENT_ITEM_TAXONOMY_ELEMENT__/g,
        getKontentItemElementTypeNameByType('taxonomy', config),
      )
      .replace(
        /__KONTENT_ITEM_URL_SLUG_ELEMENT__/g,
        getKontentItemElementTypeNameByType('url_slug', config),
      )
      // element values
      .replace(
        /__KONTENT_ELEMENT_MULTIPLE_CHOICE_VALUE__/g,
        getKontentItemElementValueTypeNameByType('multiple_choice', config),
      )
      .replace(
        /__KONTENT_ELEMENT_ASSET_VALUE__/g,
        getKontentItemElementValueTypeNameByType('asset', config),
      )
      .replace(
        /__KONTENT_ELEMENT_TAXONOMY_VALUE__/g,
        getKontentItemElementValueTypeNameByType('taxonomy', config),
      )
      .replace(
        /__KONTENT_ELEMENT_RICH_TEXT_IMAGE_VALUE__/g,
        `${getKontentItemElementValueTypeNameByType(
          RICH_TEXT_ELEMENT_TYPE_NAME,
          config,
        )}${CONNECTOR}image`,
      )
      .replace(
        /__KONTENT_ELEMENT_RICH_TEXT_LINK_VALUE__/g,
        `${getKontentItemElementValueTypeNameByType(
          RICH_TEXT_ELEMENT_TYPE_NAME,
          config,
        )}${CONNECTOR}link`,
      )
      // extensions
      .replace(
        /__KONTENT_ITEM_LANGUAGE_EXTENSION__/g,
        getKontentItemLanguageLinkExtensionName(config),
      )
      .replace(
        /__KONTENT_ITEM_PREFERRED_LANGUAGE_IDENTIFIER__/g,
        PREFERRED_LANGUAGE_IDENTIFIER,
      )
  );
};

const getKontentTaxonomiesSchemaNamingConfiguration = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string => {
  return TaxonomiesSchemaTemplate
    .replace(/__KONTENT_TAXONOMY_NAME__/g, getKontentTaxonomyTypeName(config))
    .replace(
      /__KONTENT_TAXONOMY_SYSTEM_TYPE__/g,
      getKontentTaxonomySystemElementTypeName(config),
    )
    .replace(
      /__KONTENT_TAXONOMY_TERM_TYPE__/g,
      getKontentTaxonomyTermTypeName(config),
    );
};

const getKontentTypesSchemaNamingConfiguration = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string => {
  return TypesSchemaTemplate
    .replace(/__KONTENT_TYPE_NAME__/g, getKontentTypeTypeName(config))
    .replace(
      /__KONTENT_TYPE_SYSTEM_TYPE__/g,
      getKontentTypeSystemElementTypeName(config),
    )
    .replace(
      /__KONTENT_TYPE_ELEMENT_TYPE__/g,
      getKontentTypeElementTypeName(config),
    )
    .replace(
      /__KONTENT_TYPE_ELEMENT_OPTIONS_TYPE__/g,
      getKontentTypeElementOptionTypeName(config),
    );
};

export {
  getKontentItemNodeStringForId,
  getKontentItemNodeTypeName,
  getKontentItemSystemElementTypeName,
  getKontentItemInterfaceName,
  getKontentItemElementTypeNameByType,
  getKontentItemElementsSchemaTypeName,
  getKontentItemLanguageLinkExtensionName,
  getKontentTaxonomyNodeStringForCodeName,
  getKontentTaxonomyTypeName,
  getKontentTypeNodeStringForCodeName,
  getKontentTypeTypeName,
  getKontentTypesCacheKey,
  getKontentItemsSchemaNamingConfiguration,
  getKontentTaxonomiesSchemaNamingConfiguration,
  getKontentTypesSchemaNamingConfiguration,
  PREFERRED_LANGUAGE_IDENTIFIER,
  RICH_TEXT_ELEMENT_TYPE_NAME
};
