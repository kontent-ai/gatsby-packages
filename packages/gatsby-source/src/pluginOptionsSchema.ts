
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const pluginOptionsSchema = ({ Joi }: { Joi: any }): any => {
  return Joi.object({
    projectId: Joi.string()
      .required()
      .length(36)
      .description(`Project ID from "Project settings" -> "API keys".`),
    languageCodenames: Joi.array().items(Joi.string())
      .required()
      .min(1)
      .description(`Array of language codenames from "Project settings" -> "Localization" - the first one is considered as the default one.`),
    includeTypes: Joi.boolean()
      .description(`Include content type nodes to GraphQL model.`)
      .default(false),
    includeTaxonomies: Joi.boolean()
      .description(`Include taxonomy node to GraphQL model.`)
      .default(false),
    authorizationKey: Joi.string()
      .description(`For preview/secured API key - depends on "usePreviewUrl" config.`),
    usePreviewUrl: Joi.boolean()
      .description(`When true, "preview-deliver.kontent.ai" used as primary domain for data source.`)
      .default(false),
    proxy: Joi.object()
      .keys({
        deliveryDomain: Joi.string()
          .description(`Base url used for all requests. Defaults to "deliver.kontent.ai".`),
        previewDeliveryDomain: Joi.string()
          .description(`Base url used for preview requests. Defaults to "preview-deliver.kontent.ai".`)
      })
      .description(`Proxy domain setting object`),
    includeRawContent: Joi.boolean()
      .description("Include `internal.content` property as a part fo the GraphQL model.")
      .default(false),
    experimental: Joi.object()
    .keys({
      managementApiTriggersUpdate: Joi.boolean()
      .description("Turn on experimental handling of management API triggers for update.")
      .default(false),
      additionalItemFilterParams : Joi.string()
      .description("Additional item filter parameters to reduce the content in GraphQL model.")
      .default("")
    })
  });
}

export {
  pluginOptionsSchema
}