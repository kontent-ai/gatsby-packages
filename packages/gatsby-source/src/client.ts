import * as rax from 'retry-axios';
import axios, { AxiosError } from 'axios';
import {
  KontentItemInput,
  KontentType,
  KontentTaxonomy,
  CustomPluginOptions,
} from './types';
import * as _ from 'lodash';
import {
  name as packageName,
  version as packageVersion,
} from '../package.json';
import { GatsbyCache } from 'gatsby';
import { getKontentTypesCacheKey } from './naming';

const DefaultKontentDeliveryProductionDomain = 'deliver.kontent.ai';
const DefaultKontentDeliveryPreviewDomain = 'preview-deliver.kontent.ai';
const continuationHeaderName = 'x-continuation';
const authorizationHeaderName = 'authorization';
const trackingHeaderName = 'x-kc-source';
const waitForLoadingNewContentHeaderName = 'x-kc-wait-for-loading-new-content';

rax.attach();

const getProtocolAndDomain = (options: CustomPluginOptions): string => {
  const domain = options.usePreviewUrl
    ? options?.proxy?.previewDeliveryDomain || DefaultKontentDeliveryPreviewDomain
    : options?.proxy?.deliveryDomain || DefaultKontentDeliveryProductionDomain;
  return `https://${domain}`;
};

const logRetryAttempt = (err: AxiosError): void => {
  const cfg = rax.getConfig(err);
  
  console.log(
    `Error axios request:(url: ${err.response?.config.url}) ${err.message}`,
  );
  console.log(`Retry attempt #${cfg?.currentRetryAttempt}`);
};

interface KontentHttpHeaders {
  [continuationHeaderName]?: string;
  [authorizationHeaderName]?: string;
  [trackingHeaderName]?: string;
  [waitForLoadingNewContentHeaderName]?: string;
}

const ensureAuthorizationHeader = (
  config: CustomPluginOptions,
  headers?: KontentHttpHeaders | undefined,
): KontentHttpHeaders => {
  if (!config.authorizationKey) {
    return headers || {};
  }

  if (headers) {
    headers[authorizationHeaderName] = `Bearer ${config.authorizationKey}`;
    return headers;
  } else {
    return {
      [authorizationHeaderName]: `Bearer ${config.authorizationKey}`,
    };
  }
};

const ensureNewContentHeader = (
  headers?: KontentHttpHeaders | undefined,
): KontentHttpHeaders => {
  const headerValue = `true`;
  if (headers) {
    headers[waitForLoadingNewContentHeaderName] = headerValue;
    return headers;
  } else {
    return {
      [waitForLoadingNewContentHeaderName]: headerValue,
    };
  }
};

const ensureTrackingHeader = (
  headers?: KontentHttpHeaders | undefined,
): KontentHttpHeaders => {
  const headerValue = `${packageName};${packageVersion}`;
  if (headers) {
    headers[trackingHeaderName] = headerValue;
    return headers;
  } else {
    return {
      [trackingHeaderName]: headerValue,
    };
  }
}

const loadAllKontentItems = async (
  config: CustomPluginOptions,
  language: string,
): Promise<KontentItemInput[]> => {
  let continuationToken = '';
  const items = [];
  let headers = ensureAuthorizationHeader(config);
  headers = ensureTrackingHeader(headers);
  do {
    headers[continuationHeaderName] = continuationToken;

    const response = await axios.get(
      `${getProtocolAndDomain(config)}/${
        config.projectId
      }/items-feed?language=${language}${config?.experimental?.additionalItemFilterParams ? '&' + config.experimental.additionalItemFilterParams : ''}`,
      {
        headers,
        raxConfig: {
          onRetryAttempt: logRetryAttempt,
        },
      },
    );

    const union = _.unionBy<KontentItemInput>(
      response.data.items,
      Object.values(response.data.modular_content),
      'system.codename',
    );
    items.push(...union);
    continuationToken = response.headers[continuationHeaderName];
  } while (continuationToken);

  return items;
};

const loadKontentItem = async (
  itemId: string,
  language: string,
  config: CustomPluginOptions,
  waitForLoadingNewContent = false,
): Promise<{
  item: KontentItemInput | undefined;
  modularKontent: { [key: string]: KontentItemInput };
}> => {

  let headers = ensureAuthorizationHeader(config);
  headers = ensureTrackingHeader(headers);
  if (waitForLoadingNewContent) {
    headers = ensureNewContentHeader(headers)
  }

  const response = await axios.get(
    `${getProtocolAndDomain(config)}/${config.projectId
    }/items?system.id=${itemId}&language=${language}${config?.experimental?.additionalItemFilterParams ? '&' + config.experimental.additionalItemFilterParams : ''}`,
    {
      headers,
      raxConfig: {
        onRetryAttempt: logRetryAttempt,
      },
    },
  );

  return {
    item: response.data.items.length > 0 ? response.data.items[0] : undefined,
    modularKontent: response.data.modular_content
  };
};

const loadAllKontentTypes = async (
  config: CustomPluginOptions,
): Promise<KontentType[]> => {
  let headers = ensureAuthorizationHeader(config);
  headers = ensureTrackingHeader(headers);
  const response = await axios.get(
    `${getProtocolAndDomain(config)}/${config.projectId}/types`,
    {
      headers,
      raxConfig: {
        onRetryAttempt: logRetryAttempt,
      },
    },
  );
  return response.data.types;
};

const loadAllKontentTypesCached = async (
  config: CustomPluginOptions,
  cache: GatsbyCache
): Promise<KontentType[]> => {
  let types = await cache.get(getKontentTypesCacheKey());
  if (!types) {
    types = await loadAllKontentTypes(config);
    cache.set(getKontentTypesCacheKey(), types)
  }

  return types;
}

const loadAllKontentTaxonomies = async (
  config: CustomPluginOptions,
): Promise<KontentTaxonomy[]> => {
  let headers = ensureAuthorizationHeader(config);
  headers = ensureTrackingHeader(headers);
  const response = await axios.get(
    `${getProtocolAndDomain(config)}/${config.projectId}/taxonomies`,
    {
      headers,
      raxConfig: {
        onRetryAttempt: logRetryAttempt,
      },
    },
  );
  return response.data.taxonomies;
};

export { loadKontentItem, loadAllKontentItems, loadAllKontentTypes, loadAllKontentTypesCached, loadAllKontentTaxonomies };
