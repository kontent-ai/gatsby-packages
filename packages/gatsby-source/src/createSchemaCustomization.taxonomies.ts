import { CreateSchemaCustomizationArgs } from 'gatsby';
import { getKontentTaxonomiesSchemaNamingConfiguration } from './naming';

const createSchemaCustomization = async (
  api: CreateSchemaCustomizationArgs,
): Promise<void> => {
  const schema = getKontentTaxonomiesSchemaNamingConfiguration();
  api.actions.createTypes(schema);
};

export { createSchemaCustomization as kontentTaxonomiesCreateSchemaCustomization };
