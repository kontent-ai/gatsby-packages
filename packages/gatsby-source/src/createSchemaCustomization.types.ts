import { CreateSchemaCustomizationArgs } from 'gatsby';
import { getKontentTypesSchemaNamingConfiguration } from './naming';

const createSchemaCustomization = async (
  api: CreateSchemaCustomizationArgs,
): Promise<void> => {
  const schema = getKontentTypesSchemaNamingConfiguration();
  api.actions.createTypes(schema);
};

export { createSchemaCustomization as kontentTypesCreateSchemaCustomization };
