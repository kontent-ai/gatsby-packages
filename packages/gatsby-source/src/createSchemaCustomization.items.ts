import { CreateSchemaCustomizationArgs } from 'gatsby';
import {
  CustomPluginOptions,
  KontentItemInput,
  KontentType,
  KontentTypeElementsObject,
  KontentTypeItemObjectDefinition,
} from './types';
import { loadAllKontentTypesCached } from './client';

import {
  getKontentItemElementsSchemaTypeName,
  getKontentItemNodeTypeName,
  getKontentItemsSchemaNamingConfiguration,
  getKontentItemSystemElementTypeName,
  getKontentItemInterfaceName,
  getKontentItemElementTypeNameByType,
  getKontentItemLanguageLinkExtensionName,
  PREFERRED_LANGUAGE_IDENTIFIER,
} from './naming';

const getLanguageLinkExtension = (): object => ({
  name: getKontentItemLanguageLinkExtensionName(),
  extend: (): object => ({
    async resolve(
      source: { value?: string[]; modular_content?: string[]; type: string },
      _args: unknown,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      context: any,
    ): Promise<KontentItemInput[]> {
      const nodesCodeNames =
        source.type === 'modular_content'
          ? (source.value as string[])
          : (source.modular_content as string[]);

      if (nodesCodeNames.length === 0) {
        return [];
      }

      const kontentItemNode = context.nodeModel.findRootNodeAncestor(source);
      const nodesLanguage = kontentItemNode[PREFERRED_LANGUAGE_IDENTIFIER];

      const promises = nodesCodeNames.map(codename =>
        context.nodeModel.findOne({
          query: {
            filter: {
              system: {
                codename: {
                  eq: codename,
                },
              },
              [PREFERRED_LANGUAGE_IDENTIFIER]: {
                eq: nodesLanguage,
              },
            },
          },
          type: getKontentItemInterfaceName(),
        }),
      );

      const nodes = await Promise.all(promises);
      return nodes;
    },
  }),
});

const getElementFieldsDefinitionForType = (
  type: KontentType,
): { [key: string]: { type: string } } => {
  const elementFields: {
    [key: string]: {
      type: string;
    };
  } = {};

  for (const elementKey in type.elements) {
    if (Object.prototype.hasOwnProperty.call(type.elements, elementKey)) {
      const element = (type.elements as KontentTypeElementsObject)[elementKey];
      const elementType = getKontentItemElementTypeNameByType(element.type);
      if (elementType !== '') {
        elementFields[elementKey] = {
          type: elementType,
        };
      }
    }
  }

  return elementFields;
};

const createSchemaCustomization = async (
  api: CreateSchemaCustomizationArgs,
  pluginConfig: CustomPluginOptions,
): Promise<void> => {
  // TODO check https://github.com/gatsbyjs/gatsby/pull/14610/files/5c50c435ab49884b6d854cd07f20efd95d1e5f52#diff-29de3acf9ce1010435f2b2f0043dba8cR252
  // failing for update run
  const languageExtension = getLanguageLinkExtension();
  api.actions.createFieldExtension(languageExtension);

  const baseSchemaTypes = getKontentItemsSchemaNamingConfiguration();
  api.actions.createTypes(baseSchemaTypes);

  const types = await loadAllKontentTypesCached(pluginConfig, api.cache);
  for (const type of types) {
    const kontentItemElementsTypeName = getKontentItemElementsSchemaTypeName(
      type.system.codename,
    );

    const elementsFieldDefinition = getElementFieldsDefinitionForType(type)
    const typeContainsElements = Object.keys(elementsFieldDefinition).length > 0;
    if (typeContainsElements) {
      const elementsTypeDef = api.schema.buildObjectType({
        name: kontentItemElementsTypeName,
        fields: elementsFieldDefinition,
        extensions: {
          infer: false
        }
      });
      api.actions.createTypes(elementsTypeDef);
    }

    const typeName = getKontentItemNodeTypeName(type.system.codename);
    const systemElementsTypeName = getKontentItemSystemElementTypeName();
    const typeInterfaceName = getKontentItemInterfaceName();
    const typeItemObjectDefinition: KontentTypeItemObjectDefinition = {
      name: typeName,
      fields: {
        system: `${systemElementsTypeName}!`,
        elements: kontentItemElementsTypeName,
        [PREFERRED_LANGUAGE_IDENTIFIER]: 'String!',
      },
      interfaces: ['Node', typeInterfaceName],
      infer: false,
    };
    if (!typeContainsElements) {
      delete typeItemObjectDefinition.fields.elements;
    }
    const typeItemDef = api.schema.buildObjectType(typeItemObjectDefinition);
    api.actions.createTypes(typeItemDef);
  }
};

export { createSchemaCustomization as kontentItemsCreateSchemaCustomization };
