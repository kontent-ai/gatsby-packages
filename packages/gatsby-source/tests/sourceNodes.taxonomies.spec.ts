import { kontentTaxonomiesSourceNodes } from '../src/sourceNodes.taxonomies';
import { SourceNodesArgs, Actions } from 'gatsby';
import { CustomPluginOptions, KontentTaxonomy } from '../src/types';
import { createMock } from 'ts-auto-mock';
import { createContentDigest } from 'gatsby-core-utils';

// TODO fix lint error
import * as complexTaxonomiesFakeResponse from './complexTaxonomiesFakeResponse.json';

import { mocked } from 'ts-jest/dist/util/testing';
import * as _ from 'lodash';
import axios from 'axios';
jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('sourceNodes', () => {
  describe('complex taxonomies', () => {
    const api = createMock<SourceNodesArgs>({
      createNodeId: jest.fn(input => `dummyId-${input}`),
      actions: createMock<Actions>({
        createNode: jest.fn(),
      }),
      createContentDigest,
    });

    const pluginConfiguration = createMock<CustomPluginOptions>({
      projectId: 'dummyProject',
    });

    mockedAxios.get.mockImplementation(url => {
      if (url.includes('taxonomies')) {
        return Promise.resolve({
          data: complexTaxonomiesFakeResponse,
          headers: [],
        });
      } else {
        throw new Error('Language should be defined in the language parameter');
      }
    });

    it('import all taxonomies correctly', async () => {
      await kontentTaxonomiesSourceNodes(api, pluginConfiguration);
      const createNodesMock = mocked(api.actions.createNode, true);
      const createdNodes = _.flatMap(
        createNodesMock.mock.calls,
      ) as KontentTaxonomy[];
      expect(createdNodes.length).toBe(5);
      expect(createdNodes).toMatchSnapshot();
    });
  });
});
