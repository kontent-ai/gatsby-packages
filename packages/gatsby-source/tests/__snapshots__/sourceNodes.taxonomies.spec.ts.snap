// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sourceNodes complex taxonomies import all taxonomies correctly 1`] = `
Array [
  Object {
    "children": Array [],
    "id": "dummyId-kontent_taxonomy_manufacturer",
    "internal": Object {
      "contentDigest": "33cb7b1e3bb8cdd23714f73cc2a9582f",
      "type": "kontent_taxonomy",
    },
    "system": Object {
      "codename": "manufacturer",
      "id": "4ce421e9-c403-eee8-fdc2-74f09392a749",
      "last_modified": "2017-09-07T08:15:22.7210000Z",
      "name": "Manufacturer",
    },
    "terms": Array [
      Object {
        "codename": "aerobie",
        "name": "Aerobie",
        "terms": Array [],
      },
      Object {
        "codename": "chemex",
        "name": "Chemex",
        "terms": Array [],
      },
      Object {
        "codename": "espro",
        "name": "Espro",
        "terms": Array [],
      },
      Object {
        "codename": "hario",
        "name": "Hario",
        "terms": Array [],
      },
    ],
  },
  Object {
    "children": Array [],
    "id": "dummyId-kontent_taxonomy_personas",
    "internal": Object {
      "contentDigest": "f7d8572e273960df2d6c25b68ea69fe9",
      "type": "kontent_taxonomy",
    },
    "system": Object {
      "codename": "personas",
      "id": "f30c7f72-e9ab-8832-2a57-62944a038809",
      "last_modified": "2016-10-20T13:24:00.3200000Z",
      "name": "Personas",
    },
    "terms": Array [
      Object {
        "codename": "coffee_expert",
        "name": "Coffee expert",
        "terms": Array [
          Object {
            "codename": "barista",
            "name": "Barista",
            "terms": Array [],
          },
          Object {
            "codename": "cafe_owner",
            "name": "Cafe owner",
            "terms": Array [],
          },
        ],
      },
      Object {
        "codename": "coffee_enthusiast",
        "name": "Coffee enthusiast",
        "terms": Array [
          Object {
            "codename": "coffee_lover",
            "name": "Coffee lover",
            "terms": Array [],
          },
          Object {
            "codename": "coffee_blogger",
            "name": "Coffee blogger",
            "terms": Array [],
          },
        ],
      },
    ],
  },
  Object {
    "children": Array [],
    "id": "dummyId-kontent_taxonomy_processing",
    "internal": Object {
      "contentDigest": "268a49e33cd58cacd9aabec5591e01e1",
      "type": "kontent_taxonomy",
    },
    "system": Object {
      "codename": "processing",
      "id": "d351400e-0290-87b2-1413-6c411d8ae5a4",
      "last_modified": "2017-09-07T08:15:33.7560000Z",
      "name": "Processing",
    },
    "terms": Array [
      Object {
        "codename": "wet__washed_",
        "name": "Wet (Washed)",
        "terms": Array [],
      },
      Object {
        "codename": "dry__natural_",
        "name": "Dry (Natural)",
        "terms": Array [],
      },
      Object {
        "codename": "semi_dry",
        "name": "Semi-dry",
        "terms": Array [],
      },
    ],
  },
  Object {
    "children": Array [],
    "id": "dummyId-kontent_taxonomy_product_status",
    "internal": Object {
      "contentDigest": "39ecca99c024ae9e49de57a95d394192",
      "type": "kontent_taxonomy",
    },
    "system": Object {
      "codename": "product_status",
      "id": "79b1c5b6-30bc-d076-a236-d9ec9f1ff01b",
      "last_modified": "2016-09-15T10:53:25.2230000Z",
      "name": "Product status",
    },
    "terms": Array [
      Object {
        "codename": "on_sale",
        "name": "On sale",
        "terms": Array [],
      },
      Object {
        "codename": "bestseller",
        "name": "Bestseller",
        "terms": Array [],
      },
    ],
  },
  Object {
    "children": Array [],
    "id": "dummyId-kontent_taxonomy_sitemap_538125f",
    "internal": Object {
      "contentDigest": "9c468c54eb8d79265cef978b3cbc818b",
      "type": "kontent_taxonomy",
    },
    "system": Object {
      "codename": "sitemap_538125f",
      "id": "538125ff-c59c-4193-88ce-71a7016830ed",
      "last_modified": "2019-03-27T13:02:50.1430000Z",
      "name": "Sitemap",
    },
    "terms": Array [
      Object {
        "codename": "home",
        "name": "Home",
        "terms": Array [],
      },
      Object {
        "codename": "products",
        "name": "Products",
        "terms": Array [
          Object {
            "codename": "coffee",
            "name": "Coffee",
            "terms": Array [],
          },
          Object {
            "codename": "brewers",
            "name": "Brewers",
            "terms": Array [],
          },
          Object {
            "codename": "accessories",
            "name": "Accessories",
            "terms": Array [],
          },
          Object {
            "codename": "grinders",
            "name": "Grinders",
            "terms": Array [],
          },
        ],
      },
      Object {
        "codename": "cafes",
        "name": "Cafes",
        "terms": Array [
          Object {
            "codename": "north_america",
            "name": "North America",
            "terms": Array [],
          },
          Object {
            "codename": "europe",
            "name": "Europe",
            "terms": Array [],
          },
          Object {
            "codename": "australia",
            "name": "Australia",
            "terms": Array [],
          },
        ],
      },
      Object {
        "codename": "articles",
        "name": "Articles",
        "terms": Array [],
      },
      Object {
        "codename": "offices",
        "name": "Offices",
        "terms": Array [],
      },
      Object {
        "codename": "about_us",
        "name": "About us",
        "terms": Array [],
      },
    ],
  },
]
`;
